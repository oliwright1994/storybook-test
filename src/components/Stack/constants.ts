export enum StackElement {
  div = 'div',
  ol = 'ol',
  ul = 'ul',
}

export enum StackItemElement {
  div = 'div',
  li = 'li',
}
