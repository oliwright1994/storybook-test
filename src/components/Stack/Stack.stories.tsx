import { Text } from '@chakra-ui/core'
import { boolean, select } from '@storybook/addon-knobs'
import React from 'react'

import { HorizontalAlignment, Space } from '../../constants'
import createStory from '../../storybook/createStory'
import Stack from './Stack'

createStory('COMPONENTS | Stack', [
  {
    label: 'Stack',
    jsx: () => {
      const space = select('Space', Space, Space.s)
      const alignItems = select('Align', HorizontalAlignment, HorizontalAlignment.left)
      const dividers = boolean('Dividers', true)

      return (
        <Stack space={space} alignItems={alignItems} dividers={dividers}>
          <Text>Test one</Text>
          <Text>Test two</Text>
          <Text>Test Three</Text>
        </Stack>
      )
    },
  },
])
