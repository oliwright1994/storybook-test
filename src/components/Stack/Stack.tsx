import { Box, Flex } from '@chakra-ui/core'
import React, { FC } from 'react'

import { HorizontalAlignment, Space } from '../../constants'
import { StackElement, StackItemElement } from './constants'

interface IStack {
  as?: StackElement
  space?: Space
  alignItems?: HorizontalAlignment
  dividers?: boolean
}
interface IStackItem {
  as: StackItemElement
  space?: Space
  alignItems: HorizontalAlignment
}

const alignItemsMap = {
  left: 'flex-start',
  center: 'center',
  right: 'flex-end',
}

const StackItem: FC<IStackItem> = ({ alignItems, space, ...props }) => (
  <Flex direction="column" alignItems={alignItemsMap[alignItems]} mt={space} {...props} />
)

const Stack: FC<IStack> = ({
  children,
  as = StackElement.div,
  space = Space.none,
  alignItems = HorizontalAlignment.left,
  dividers,
}) => {
  const isList = as === StackElement.ol || as === StackElement.ul
  const stackItemElement = isList ? StackItemElement.div : StackItemElement.li
  const stackItemProps: IStackItem = {
    as: stackItemElement,
    alignItems,
  }

  return (
    <Box as={as}>
      {React.Children.map(children, (child, index) =>
        index === 0 ? (
          <StackItem {...stackItemProps}>{child}</StackItem>
        ) : (
          <>
            {dividers && <Box as="hr" mt={space} />}
            <StackItem {...stackItemProps} space={space}>
              {child}
            </StackItem>
          </>
        )
      )}
    </Box>
  )
}

export default Stack
