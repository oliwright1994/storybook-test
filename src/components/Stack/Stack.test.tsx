import { Text } from '@chakra-ui/core'
import { screen } from '@testing-library/react'
import React from 'react'

import { renderWithBrightTheme } from '../../../testing/helpers/themeHelper'
import Stack from '.'

describe('<Stack />', () => {
  it('should render as expected', () => {
    const { container } = renderWithBrightTheme(
      <Stack>
        <Text>Test One</Text>
        <Text>Test Two</Text>
        <Text>Test Three</Text>
      </Stack>
    )
    const textNodes = screen.getAllByText(/test/i)

    expect(textNodes).toHaveLength(3)
    expect(container).toMatchSnapshot()
  })
})
