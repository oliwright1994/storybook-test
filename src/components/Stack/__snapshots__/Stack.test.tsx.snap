// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<Stack /> should render as expected 1`] = `
<div>
  <div
    class="css-0"
  >
    <li
      class="css-1lxwves"
    >
      <p
        class="css-fzcsno"
      >
        Test One
      </p>
    </li>
    <li
      class="css-kuxuyf"
    >
      <p
        class="css-fzcsno"
      >
        Test Two
      </p>
    </li>
    <li
      class="css-kuxuyf"
    >
      <p
        class="css-fzcsno"
      >
        Test Three
      </p>
    </li>
  </div>
</div>
`;
